### DISCUSSION TITLE
TypeVar "T" appears only once in generic function signature

### DISCUSSION THREAD

**Discussion by bitsandscraps** [microsoft/pyright#7213] [Q&A] [answered]
The code
```python
from typing import TypeAlias, TypeVar


T = TypeVar('T')
A: TypeAlias = list[T]


def append(a: A, b: T) -> A:
    return a + [b]
```
results in the following warning.
```
warning: TypeVar "T" appears only once in generic function signature
Use "object" instead (reportInvalidTypeVarUse)
```
Is aliasing type variables not recommended?

**Comment by erictraut** [comment, marked as answer]
You are not aliasing a type variable here. You are defining a generic type alias `A` with a single type parameter `T`. When you use the type alias `A`, you should supply a single type argument. Since you have omitted the type argument, a type checker will assume that you meant `Any`. That means your function is effectively:

```python
def append(a: A[Any], b: T) -> A[Any]: ...
```

I assume that what you intended is:

```python
def append(a: A[T], b: T) -> A[T]: ...
```

You're using the old way to define type variables and generics. Starting with Python 3.12, you can use the [PEP 695](https://peps.python.org/pep-0695/) syntax, which improves the usability of generics. Here's what your code would look like using the new syntax.

```python
type A[T] = list[T]

def append[T](a: A[T], b: T) -> A[T]:
    return a + [b]
```

**Reply by bitsandscraps** [reply]
I didn't know that a type becomes generic when you're aliasing it. Thank you so much! I'll also try the new syntax when I get to use Python 3.12.
